### PR TITLE
feat(cspell): add self healing capability

### PR DESCRIPTION
**Pull request by mfranzke**
## Proposed changes

added self healing capability for `cspell`.

### Code Quality

- [x] I have reviewed my own code (self-review), including the screen- and snapshots
- [x] I have reviewed my changes locally with an AI assistant (e.g. GitHub Copilot, Kiro, etc.)
- [x] No hardcoded values, magic numbers, or debug code left in

### Validation

- [x] I have added/updated tests that cover my changes
- [ ] ~I have tested across all relevant frameworks (React, Angular, Vue) if applicable~

### General

- [x] The PR title follows the conventional commits format (e.g. `feat:`, `fix:`, `chore:`), as in [Git Commit Conventions](docs/conventions.md#git-commits-conventions)
- [ ] If architecture, structure, or conventions changed inside a `packages/*` folder, the corresponding `AGENTS.md` has been updated

🔭🐙🐈 Test this branch here: <https://design-system.deutschebahn.com/core-web/review/feat-cspell-add-self-healing-capability>

### CODE REVIEW

File: scripts/cspell-add-igo-words.mjs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+async function main() {
+	try {
+		await execFileAsync('pnpm', ['run', 'lint:codespell']);
+	} catch (error) {
+		const output = `${error.stdout ?? ''}\n${error.stderr ?? ''}`;
+
+		const words = [...output.matchAll(/Unknown word\s+\(([^)]+)\)/g)].map(
+			(match) => match[1]
+		);
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply scripts/cspell-add-igo-words.mjs]: @copilot please provide a test as suggested.

File: .config/cspellignorewords.txt [hunk 2]
```diff
@@ -1,205 +1,204 @@
...
+michaelmkraus # Username
+minireset
+modelcontextprotocol
+moz
+mypath
+navigations
+navs
+networkidle
+ngmodel
+nmerget # Username
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply .config/cspellignorewords.txt:119]: hardened our script instead.

File: .config/cspellignorewords.txt [hunk 3]
```diff
@@ -1,205 +1,204 @@
...
+licence
+lightningcss
+lintstagedrc
+llms
+loadingindicator
+mainnavigation
+mdfiles
+merget
+mfranzke # Username
+michaelmkraus # Username
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply .config/cspellignorewords.txt:107]: hardened our script instead.

File: scripts/cspell-add-ignore-words.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	for (const line of content.split(/\r?\n/)) {
+		const normalizedWord = normalizeIgnoreWord(line);
+
+		if (!normalizedWord) {
+			continue;
+		}
+
+		// Persist original line incl. comment
+		entries.set(normalizedWord, line);
+	}
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply scripts/cspell-add-ignore-words.ts:71]: I've corrected the test, as this was actually incorrect.

File: scripts/tests/cspell-add-ignore-words.spec.mjs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		expect(extractUnknownWords(output)).toEqual(['Foobar', 'Baz']);
+	});
+});
+
+describe('mergeIgnoreWords', () => {
+	test('trims and merges words into sorted unique output', () => {
+		const existing = '  Zebra\nAlpha # Username\n\nBeta  \n';
+		const merged = mergeIgnoreWords(existing, ['Beta', 'Gamma', '  Delta']);
+
+		expect(merged).toBe('Alpha\nBeta\nDelta\nGamma\nZebra\n');
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply scripts/tests/cspell-add-ignore-words.spec.mjs]: I've corrected the test, as this was actually incorrect.

File: scripts/cspell-add-ignore-words.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		if (words.length === 0) {
+			throw error;
+		}
+
+		await addWords(words);
+
+		console.log(
+			`Ensured ${words.length} word(s) are listed in ${IGNORE_FILE}.`
+		);
+	}
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply scripts/cspell-add-ignore-words.ts:120]: we should keep this simple, as the purpose of this is only adding the words, and not taking into account any errors. All related processes (either lint-staged or the pipeline) would re-run anyhow.

File: scripts/cspell-add-ignore-words.ts
Comment (nmerget) [inline scripts/cspell-add-ignore-words.ts:1]: We shall use `.ts` files if there is no requirement to use JS

**Comment by nmerget** [comment]
## AI comments:

### 3. The `isDirectExecution` guard is fragile

```js
const isDirectExecution = process.argv.some(
    (arg) => arg !== undefined && path.resolve(arg) === selfPath
);
```

This works for `node scripts/cspell-add-ignore-words.mjs` but may not match if invoked via an absolute path with different casing (Windows) or via symlinks. A simpler pattern would be:

```js
const isDirectExecution =
    path.resolve(process.argv[1]) === selfPath;
```

`process.argv[1]` is always the script entry point. The `.some()` loop over all args is unnecessary and could false-positive if the script path happens to appear as an argument to another script.

### 4. Locale-sensitive sorting

```js
.toSorted(([a], [b]) => a.localeCompare(b))
```

`localeCompare` without an explicit locale means the sort order depends on the runner's system locale. For a committed file that should be deterministic, consider:

```js
a.localeCompare(b, 'en', { sensitivity: 'base' })
```

or use simple ASCII comparison since these are code identifiers:

```js
a.toLowerCase().localeCompare(b.toLowerCase())
```

### 5. Case sensitivity on deduplication

`entries.has(normalizedWord)` is case-sensitive. If the ignore-words list already contains `Foobar` and cspell reports `foobar`, the script will add it as a separate entry. This might be intentional (cspell is case-sensitive), but worth confirming.

### 7. Missing edge-case test: empty input to `extractUnknownWords`

The test suite doesn't cover `extractUnknownWords('')` or `extractUnknownWords('no matches here')`. Adding a quick assertion would strengthen confidence.

### 8. `main()` silently succeeds when cspell passes

If `lint:codespell` passes (no errors), the script exits silently without any feedback. A `console.log('No spelling issues found.')` in the try-block would make the script more user-friendly when run manually.

**Comment by mfranzke** [comment]
> ## AI comments:
> ### 5. Case sensitivity on deduplication
> `entries.has(normalizedWord)` is case-sensitive. If the ignore-words list already contains `Foobar` and cspell reports `foobar`, the script will add it as a separate entry. This might be intentional (cspell is case-sensitive), but worth confirming.

I even also already wondered about case-sensitive entries in the list, but as cspell handles those case-sensitive, it's not a problem, but needed to do it this way.

File: scripts/cspell-add-ignore-words.ts
Comment (mfranzke) [reply scripts/cspell-add-ignore-words.ts:1]: @copilot please take (even also typing the file correctly).

File: package.json [hunk 7]
```diff
@@ -58,6 +58,7 @@
     "lint-docs-file-references": "node scripts/check-docs.js",
     "lint-packages": "pnpm tsx scripts/lint-packages.ts publint",
     "lint:codespell": "cspell . --config .config/cspell.config.ts --quiet",
+    "lint:codespell:fix": "node scripts/cspell-add-ignore-words.ts",
```
[the remark this answers is not in the document]
Comment (mfranzke) [reply package.json:61]: it does not fail in node.

File: scripts/tests/cspell-add-ignore-words.spec.mjs
Comment (nmerget) [inline scripts/tests/cspell-add-ignore-words.spec.mjs:1]: Why is this an `mjs` file as well?

File: package.json (hunk 7 above)
Comment (nmerget) [inline package.json:61]: We don't run `lint:*` with the `--fix` options here. The `pnpm run lint` should fail with exit code 1.

Instead we should run it only inside the `self-healing-orchestrator.ts` so we can move it to the `scripts/package.json`. Also we should add it to `lintstaged` to make sure we don't miss any word

**Comment by nmerget** [comment]
image

image


I tested the "auto-fix" and I'm pretty sure that self-healing isn't working as expected. It doesn't find `BLubber` but the missspelled `devloping`. This is not a `fix`. It just adds a wrong word to the ignore list. So the linter will be valid but the word is still false for documentation. 

I wouldn't reccomend to implement it.

And *if* we do it we need to do this:


## Proposed changes

Refactors `cspell-add-ignore-words.ts` to use the programmatic `cspell` API instead of shelling out, and reorganizes the fix script to only run inside the self-healing orchestrator and lintstaged.

### What changed

- **`scripts/cspell-add-ignore-words.ts`** — Replaced `execSync('pnpm run lint:codespell')` + regex output parsing with cspell's programmatic `lint()` function and a typed reporter. Uses absolute paths derived from `import.meta.url` so it works regardless of CWD.
- **`package.json` (root)** — Removed `lint:codespell:fix`. `pnpm run lint` now only checks and exits with code 1 on failure (no auto-fix at root level).
- **`scripts/package.json`** — Added `"codespell:fix": "node cspell-add-ignore-words.ts"` script and `cspell` as devDependency for the programmatic API.
- **`scripts/self-healing-orchestrator.ts`** — Updated fix command from `pnpm run lint:codespell:fix` to `pnpm --filter scripts run codespell:fix`.
- **`.config/.lintstagedrc.js`** — Extended the `'*'` entry to also run `codespell:fix` after the cspell check, so unknown words are auto-added to the ignore file during commits.
- **`scripts/tests/cspell-add-ignore-words.spec.mjs` → `.spec.ts`** — Converted test file from `.mjs` to `.ts` with proper typing.

### Why

- Parsing CLI stdout with regex is fragile and breaks when output format changes
- `execFile` / `execSync` required platform-specific workarounds (`.cmd` on Windows)
- The fix script should not be exposed as a `lint:*` command — it's only for self-healing CI and lintstaged

**Comment by mfranzke** [comment]
> I tested the "auto-fix" and I'm pretty sure that self-healing isn't working as expected. It doesn't find BLubber but the missspelled devloping. This is not a fix. It just adds a wrong word to the ignore list. So the linter will be valid but the word is still false for documentation.

> I wouldn't reccomend to implement it.

It's exactly what we do with regenerating the screen- and snapshots. I'm fine with not calling it "fix" than, as it's mainly generating a new file input, but it would help generating this output for humans to double-check afterwards, like we do with screen- and snapshots.

**Comment by nmerget** [comment]
We'll create a full user story for this in the future
